src/topology/homotopy/homotopy_group.lean

```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
--/
+/-- We allow an arbitrary indexing type `N`, not just `fin n`, in the definition of `cube`, so as to
+  be able to identify facets of a cube to cubes indexed by subtypes of `N`, which is more
+  convenient than using `fin (n-1)`. -/
 @[derive [has_zero, has_one, topological_space]]
-def cube (n : ℕ) : Type := fin n → I
-local notation `I^` := cube
+def cube (N : Type*) := N → I
+local notation `I^` n := cube (fin n)
 
 namespace cube
 
-@[continuity] lemma proj_continuous (i : fin n) : continuous (λ f : I^n, f i) :=
-continuous_apply i
-
-/--
-The points of the `n`-dimensional cube with at least one projection equal to 0 or 1.
--/
-def boundary (n : ℕ) : set (I^n) := {y | ∃ i, y i = 0 ∨ y i = 1}
+instance compact_space : compact_space (cube N) :=
+by { convert pi.compact_space, intro, apply_instance }
 
-/--
-The first projection of a positive-dimensional cube.
--/
-@[simp] def head {n} : I^(n+1) → I := λ c, c 0
+instance locally_compact_space : locally_compact_space (cube N) :=
+by convert locally_compact_space.pi; intro; apply_instance
 
-@[continuity] lemma head.continuous {n} : continuous (@head n) := proj_continuous 0
+/-- The points in a cube with at least one projection equal to 0 or 1. -/
+def boundary (N) : set (cube N) := {y | ∃ i, y i = 0 ∨ y i = 1}
 
-/--
-The projection to the last `n` coordinates from an `n+1` dimensional cube.
--/
-@[simp] def tail {n} : I^(n+1) → I^n := λ c, fin.tail c
+variable {n : ℕ}
+/-- The first projection of a positive-dimensional cube. -/
+@[simps] def head : C(I^(n+1), I) := ⟨λ t, t 0, continuous_apply 0⟩
 
 instance unique_cube0 : unique (I^0) := pi.unique_of_is_empty _
 
-lemma one_char (f : I^1) : f = λ _, f 0 := by convert eq_const_of_unique f
+lemma one_char (f : I^1) : f = λ _, f 0 := eq_const_of_unique f
+
+section
+variable [decidable_eq N]
+
+/-- The forward direction of the homeomorphism
+  between the cube $I^N$ and $I × I^{N\setminus\{j\}}$. -/
+abbreviation split_at (i : N) := fun_split_at I i
+
+/-- The backward direction of the homeomorphism
+  between the cube $I^N$ and $I × I^{N\setminus\{j\}}$. -/
+abbreviation insert_at (i : N) := (fun_split_at I i).symm
+
+lemma insert_at_boundary (i : N) {t₀ : I} {t} (H : (t₀ = 0 ∨ t₀ = 1) ∨ t ∈ boundary {j // j ≠ i}) :
+  insert_at i ⟨t₀, t⟩ ∈ boundary N :=
+begin
+  obtain H | ⟨j, H⟩ := H,
+  { use i, rwa [fun_split_at_symm_apply, dif_pos rfl] },
+  { use j, rwa [fun_split_at_symm_apply, dif_neg j.prop, subtype.coe_eta] },
+end
+
+end
 
 end cube
 
-/--
-The `n`-dimensional generalized loops; functions `I^n → X` that send the boundary to the base point.
--/
-structure gen_loop (n : ℕ) (x : X) extends C(I^n, X) :=
-(boundary : ∀ y ∈ cube.boundary n, to_fun y = x)
+/-- The space of paths with both endpoints equal to a specified point `x : X`. -/
+abbreviation loop_space (X : Type*) [topological_space X] (x : X) := path x x
+local notation `Ω` := loop_space
+
+instance loop_space.inhabited : inhabited (Ω X x) := ⟨path.refl x⟩
```

Comment:
Does `by apply_instance` work here? If so, you can just delete this.

Reply:
`apply_instance` fails, but the instance isn't used anyway so I'll delete it.

Reply:
In that case we should be adding the instance for `path` directly.

Reply:
Maybe not worth modifying one more ported file. Maybe @ralvrz will add a TODO item after port?